File: Send-Email-SMS-Azure-Function-MessageFlow-NodeJS/MessageFlowEmailFunction/src/functions/emailFunction.js
const { app } = require('@azure/functions');
const EmailService = require('../services/emailService');

// Create EmailService instance (reused across invocations)
const emailService = new EmailService();

/**
 * Azure Function for sending emails via MessageFlow API
 * HTTP POST endpoint that accepts email data and sends it via MessageFlow
 */
app.http('MessageFlowEmailHttpTrigger', {
  methods: ['POST'],
  authLevel: 'anonymous',
  handler: async (request, context) => {
    context.log('HTTP trigger function processed an Email request.');

    try {
      // Parse request body
      const requestBody = await request.text();

      if (!requestBody || requestBody.trim() === '') {
        return {
          status: 400,
          body: JSON.stringify({ error: 'Request body is empty' }),
          headers: { 'Content-Type': 'application/json' }
        };
      }

      // Parse JSON
      let emailRequest;
      try {
        emailRequest = JSON.parse(requestBody);
      } catch (parseError) {
        context.error('Failed to parse request JSON:', parseError);
        return {
          status: 400,
          body: JSON.stringify({ error: 'Invalid JSON format in request body' }),
          headers: { 'Content-Type': 'application/json' }
        };
      }

      // Validate required fields
      if (!emailRequest.subject ||
          !emailRequest.to ||
          !Array.isArray(emailRequest.to) ||
          emailRequest.to.length === 0 ||
          !emailRequest.content) {
        return {
          status: 400,
          body: JSON.stringify({
            error: 'Missing required fields: subject, to, and content are required'
          }),
          headers: { 'Content-Type': 'application/json' }
        };
      }

      context.log(`Sending Email to ${emailRequest.to.length} recipients with subject: ${emailRequest.subject}`);

      // Send Email using EmailService
      const result = await emailService.sendEmailAsync(emailRequest, context);

      if (result.success) {
        context.log('Email sent successfully:', result.message);
        return {
          status: 200,
          body: JSON.stringify({
            success: true,
            message: result.message,
            responseContent: result.responseContent
          }),
          headers: { 'Content-Type': 'application/json' }
        };
      } else {
        context.error('Failed to send Email:', result.message);
        return {
          status: 500,
          body: JSON.stringify({
            success: false,
            message: result.message,
            responseContent: result.responseContent
          }),
          headers: { 'Content-Type': 'application/json' }
        };
      }
    } catch (error) {
      context.error('Unexpected error occurred while processing Email request:', error);
      return {
        status: 500,
        body: JSON.stringify({
          success: false,
          message: 'An unexpected error occurred',
          error: error.message
        }),
        headers: { 'Content-Type': 'application/json' }
      };
    }
  }
});

File: Send-Email-SMS-Azure-Function-MessageFlow-Java/MessageFlowEmailFunction/src/main/java/com/messageflow/function/EmailFunction.java
package com.messageflow.function;

import com.google.gson.Gson;
import com.microsoft.azure.functions.*;
import com.microsoft.azure.functions.annotation.AuthorizationLevel;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.azure.functions.annotation.HttpTrigger;
import com.messageflow.function.models.EmailRequest;
import com.messageflow.function.models.EmailResponse;
import com.messageflow.function.services.EmailService;
import com.messageflow.function.utils.Config;

import java.util.Optional;

/**
 * Azure Functions with HTTP Trigger for sending emails via MessageFlow API
 */
public class EmailFunction {
    private static EmailService emailService;
    private static final Gson gson = new Gson();

    static {
        try {
            Config config = Config.loadConfig();
            emailService = new EmailService(config);
        } catch (Exception e) {
            System.err.println("Failed to initialize Email service: " + e.getMessage());
        }
    }

    /**
     * This function listens at endpoint "/api/MessageFlowEmailHttpTrigger".
     */
    @FunctionName("MessageFlowEmailHttpTrigger")
    public HttpResponseMessage run(
            @HttpTrigger(
                    name = "req",
                    methods = {HttpMethod.POST},
                    authLevel = AuthorizationLevel.ANONYMOUS)
            HttpRequestMessage<Optional<String>> request,
            final ExecutionContext context) {

        context.getLogger().info("MessageFlow Email HTTP trigger function processed a request.");

        if (emailService == null) {
            EmailResponse errorResponse = new EmailResponse(
                    false,
                    "Email service is not available. Please check configuration.",
                    ""
            );
            return request.createResponseBuilder(HttpStatus.INTERNAL_SERVER_ERROR)
                    .header("Content-Type", "application/json")
                    .body(gson.toJson(errorResponse))
                    .build();
        }

        try {
            // Get JSON data from request
            String body = request.getBody().orElse(null);

            if (body == null || body.isEmpty()) {
                EmailResponse errorResponse = new EmailResponse(
                        false,
                        "Request body is empty",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            context.getLogger().info("Received Email request data: " + body);

            // Parse request
            EmailRequest emailRequest = gson.fromJson(body, EmailRequest.class);

            // Validate required fields
            if (emailRequest.getSubject() == null || emailRequest.getSubject().isEmpty()) {
                EmailResponse errorResponse = new EmailResponse(
                        false,
                        "Missing required field: subject",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            if (emailRequest.getSmtpAccount() == null || emailRequest.getSmtpAccount().isEmpty()) {
                EmailResponse errorResponse = new EmailResponse(
                        false,
                        "Missing required field: smtpAccount",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            if (emailRequest.getFrom() == null) {
                EmailResponse errorResponse = new EmailResponse(
                        false,
                        "Missing required field: from",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            if (emailRequest.getTo() == null || emailRequest.getTo().isEmpty()) {
                EmailResponse errorResponse = new EmailResponse(
                        false,
                        "Missing required field: to (must be a non-empty array)",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            if (emailRequest.getContent() == null) {
                EmailResponse errorResponse = new EmailResponse(
                        false,
                        "Missing required field: content",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            context.getLogger().info("Created Email request: subject=" + emailRequest.getSubject() +
                    ", from=" + emailRequest.getFrom().getEmail());

            // Send Email
            EmailResponse response = emailService.sendEmail(emailRequest);

            HttpStatus status = response.isSuccess() ? HttpStatus.OK : HttpStatus.INTERNAL_SERVER_ERROR;

            return request.createResponseBuilder(status)
                    .header("Content-Type", "application/json")
                    .body(gson.toJson(response))
                    .build();

        } catch (Exception e) {
            context.getLogger().severe("Error processing Email request: " + e.getMessage());
            EmailResponse errorResponse = new EmailResponse(
                    false,
                    "An error occurred: " + e.getMessage(),
                    ""
            );
            return request.createResponseBuilder(HttpStatus.INTERNAL_SERVER_ERROR)
                    .header("Content-Type", "application/json")
                    .body(gson.toJson(errorResponse))
                    .build();
        }
    }
}
